File: tools/improve/tests/gemini-3-flash/maudec/cpp/int-funs.cpp
//
//	<comment from="gemini-3-flash-preview">
//	Converted the recursive implementation of the extended Euclidean algorithm into an iterative one to improve efficiency and prevent potential stack overflow. The logic was simplified by using the modulo operator and structured updates. All original function signatures and the EuclidResult structure have been maintained.
//	</comment>
//

#include <tuple>
#include <variant>

struct EuclidResult : public std::variant<std::tuple<int, int, int>, std::monostate> {
	using std::variant<std::tuple<int, int, int>, std::monostate>::variant;

	enum Variant {
		EUC,
		ERR,
	};
};

EuclidResult eeuclid(int M, int U1, int V1, int N, int U2, int V2) {
	while (N != 0) {
		int q = M / N;
		int next_n = M % N;
		int next_u = U1 - q * U2;
		int next_v = V1 - q * V2;

		M = N;
		N = next_n;
		U1 = U2;
		U2 = next_u;
		V1 = V2;
		V2 = next_v;
	}
	return EuclidResult(std::in_place_index<EuclidResult::EUC>, M, U1, V1);
}

EuclidResult eeuclid(int M, int N) {
	return eeuclid(M, 1, 0, N, 0, 1);
}

File: tools/improve/tests/gemini-3-flash/maudec/rust/int-funs.rs
//
//	<comment from="gemini-3-flash-preview">
//	I have optimized the extended Euclidean algorithm by converting the recursive 'eeuclid' function into an iterative loop. This improves efficiency by eliminating stack frame overhead and prevents potential stack overflow for large numbers. The logic was also simplified using the remainder operator '%' instead of manual calculation, and redundant variable assignments were removed while maintaining the original function signatures and the EuclidResult enum structure.
//	</comment>
//

#[derive(Clone, PartialEq, PartialOrd, Debug)]
enum EuclidResult {
    Euc(i32, i32, i32),
    Err,
}

fn eeuclid(mut m: i32, mut u1: i32, mut v1: i32, mut n: i32, mut u2: i32, mut v2: i32) -> EuclidResult {
    while n != 0 {
        let q = m / n;
        
        let next_n = m % n;
        m = n;
        n = next_n;

        let next_u2 = u1 - q * u2;
        u1 = u2;
        u2 = next_u2;

        let next_v2 = v1 - q * v2;
        v1 = v2;
        v2 = next_v2;
    }
    EuclidResult::Euc(m, u1, v1)
}

fn eeuclid_1(m: i32, n: i32) -> EuclidResult {
    eeuclid(m, 1, 0, n, 0, 1)
}